File: Backend/backend/agents/agent_state.py
#agent_state.py
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List

# LangGraph state model
class AgentState(BaseModel):
    messages: list = Field(default_factory=list)
    ui_chat_history: List[Dict[str, Any]] = Field(default_factory=list)
    next_node: str = "supervisor"
    iteration_count: int = 0
    dfs: dict = Field(default_factory=dict)
    df_heads: dict = Field(default_factory=dict)
    df_dtypes: dict = Field(default_factory=dict)
    dbfs_path: str = ""
    file_types: Dict[str, str] = Field(default_factory=dict)
    xml_root_tags: Optional[Dict[str, str]] = Field(default_factory=dict)

    # --- RAG mapper outputs (persist these so  can read them) ---
    mapping_rows: List[Dict[str, Any]] = Field(default_factory=list)
    fhir_mapping_rows: List[Dict[str, Any]] = Field(default_factory=list)
    gold_mapping_rows: List[Dict[str, Any]] = Field(default_factory=list)

    llm_summaries: Dict[str, str] = Field(default_factory=dict)
    col_summaries: Dict[str, Dict[str, str]] = Field(default_factory=dict)
    rag_evidence: Dict[str, Dict[str, Dict[str, List[Dict[str, Any]]]]] = Field(default_factory=dict)

    modeling_sql: str = ""
    modeling_schema_view: str = ""
    modeling_er_diagram: Dict[str, Any] = Field(default_factory=dict)

    domain: Optional[str] = None
    area: Optional[str] = None
    suggested_areas: Optional[List[str]] = None
    kpis: Optional[str] = None

    pii_columns: Optional[List[str]] = Field(default_factory=list)
    phi_columns: Optional[List[str]] = Field(default_factory=list)
    sensitive_metadata: Optional[Dict[str, Any]] = Field(default_factory=dict)

    pyspark_code: str = ""
    dq_rules: str = ""
    canvas_code: str = ""
    code_history: list = Field(default_factory=list)
    dataset_summary: Optional[Dict[str, Any]] = Field(default_factory=dict)

    # Masking agent outputs
    masking_sql: Optional[str] = None
    masking_sql_lines: Optional[List[str]] = Field(default_factory=list)
    masking_version: int = 0

    # Execution state + structured log entries
    mask_execution_status: str = "NOT_STARTED"   # "NOT_STARTED" | "RUNNING" | "SUCCESS" | "FAILED"
    mask_execution_log: List[Dict[str, Any]] = Field(default_factory=list)
    
    # Run-time / target configuration
    catalog: Optional[str] = None
    schema_name: Optional[str] = None
    pii_access_group: Optional[str] = "pii_access"
    phi_access_group: Optional[str] = "phi_access"
    target_catalog: Optional[str] = None
    target_schema: Optional[str] = None

    pii_access_mode: Optional[str] = "group"
    pii_access_value: Optional[str] = "pii_access"
    pii_access_user: Optional[str] = None
    phi_access_mode: Optional[str] = "group"
    phi_access_value: Optional[str] = "phi_access"
    phi_access_user: Optional[str] = None

    class Config:
        extra = "allow"
        arbitrary_types_allowed = True

File: Frontend/src/types/agent-state.ts
export interface AgentState {
  // Step 1 - Setup & Upload
  catalog: string;
  schema: string;
  volume: string;
  dbfs_path: string;
  dfs: Record<string, unknown>;
  df_heads: Record<string, unknown>;
  df_dtypes: Record<string, Record<string, string>>;
  file_types: Record<string, string>;
  xml_root_tags: Record<string, string>;

  // Step 3 - Data Lake Design
  mapping_rows: MappingRow[];
  fhir_mapping_rows: MappingRow[];
  gold_mapping_rows: GoldMappingRow[];
  llm_summaries: Record<string, string>;
  col_summaries: Record<string, Record<string, string>>;
  rag_evidence: Record<string, Record<string, RAGEvidence>>;

  // Step 4 - Business KPIs
  domain: string | null;
  area: string | null;
  suggested_areas: string[];
  kpis: string | null;

  // Step 5 - Code Generation
  pii_columns: string[];
  phi_columns: string[];
  sensitive_metadata: Record<string, unknown>;
  pyspark_code: string;
  dq_rules: string;
  masking_sql: string | null;
  masking_sql_lines: string[];
  mask_execution_status: 'NOT_STARTED' | 'RUNNING' | 'SUCCESS' | 'FAILED';
  mask_execution_log: MaskExecutionLog[];

  // Step 2 - Data Modeling
  modeling_sql: string;
  modeling_schema_view: string;
  modeling_er_diagram: ERDiagram;
  modeling_diagram: ERDiagramGraph;

  // Chat
  messages: Message[];
  ui_chat_history: Message[];

  // Execution config
  pii_access_mode: 'group' | 'user';
  pii_access_value: string;
  phi_access_mode: 'group' | 'user';
  phi_access_value: string;
}

export interface ERDiagram {
  tables: ERTable[];
  relationships: ERRelationship[];
}

export interface ERTable {
  name: string;
  columns: ERColumn[];
}

export interface ERColumn {
  name: string;
  data_type?: string;
  is_primary_key: boolean;
  is_foreign_key: boolean;
}

export interface ERRelationship {
  from_table: string;
  from_column: string;
  to_table: string;
  to_column: string;
  from_cardinality?: '1' | 'N';
  to_cardinality?: '1' | 'N';
}

export interface DiagramNode {
  table_name: string;
  columns: ERColumn[];
  table_type: 'dimension' | 'fact' | 'table';
}

export interface DiagramEdge {
  from_table: string;
  from_column: string;
  to_table: string;
  to_column: string;
  from_cardinality?: '1' | 'N';
  to_cardinality?: '1' | 'N';
}

export interface ERDiagramGraph {
  nodes: DiagramNode[];
  edges: DiagramEdge[];
}

export interface MappingRow {
  bronze_table: string;
  bronze_columns: string;
  silver_table: string;
  silver_column: string;
  data_classification?: 'PII' | 'PHI' | 'PCI' | 'NON_SENSITIVE';
}

export interface GoldMappingRow {
  silver_table: string;
  silver_column: string;
  gold_table: string;
  gold_column: string;
  transformation?: string;
  description?: string;
  table_type?: 'dimension' | 'fact';
}

export interface RAGEvidence {
  resource_docs: RAGDocument[];
  column_docs: RAGDocument[];
}

export interface RAGDocument {
  resource_name?: string;
  target_table?: string;
  fhir_path?: string;
  target_column?: string;
  description?: string;
  score: number | null;
  text: string;
}

export interface Message {
  role: 'user' | 'assistant' | 'system';
  content: string;
  name?: string;
  timestamp?: string;
}

export interface MaskExecutionLog {
  timestamp: string;
  level: 'info' | 'warning' | 'error';
  message: string;
}
